metrics/sinks/hawkular/driver_historical.go

```diff
@@ -0,0 +1,432 @@
+			val = bp.Max
+		case core.AggregationTypeMinimum:
+			val = bp.Min
+		case core.AggregationTypeMedian:
+			val = bp.Median
+		case core.AggregationTypePercentile50:
+			val = bp.Median // Percentile 50 and Median are the same thing
+		case core.AggregationTypePercentile95:
+			for _, v := range bp.Percentiles {
+				if v.Quantile == 95.0 {
+					val = v.Value
+				}
+			}
+		case core.AggregationTypePercentile99:
+			for _, v := range bp.Percentiles {
+				if v.Quantile == 99.0 {
+					val = v.Value
+				}
+			}
+		}
+
+		mv := core.MetricValue{
+			MetricType: core.MetricGauge,
+			ValueType:  core.ValueFloat,
+			FloatValue: float32(val),
+		}
+		f, err := metrics.ConvertToFloat64(val)
+		if err != nil {
+			return tav, err
+		}
+		mv.FloatValue = float32(f)
+
+		aggs[a] = mv
+	}
+
+	tav.AggregationValue.Aggregations = aggs
+	return tav, nil
+}
+
+func (h *hawkularSink) datapointToMetricValue(dp *metrics.Datapoint, mt *metrics.MetricType) (core.TimestampedMetricValue, error) {
+	mv := core.MetricValue{
+		MetricType: hawkularTypeToHeapsterType(*mt),
+	}
+
+	tmv := core.TimestampedMetricValue{
+		Timestamp: dp.Timestamp,
+	}
+
+	switch *mt {
+	case metrics.Counter:
+		mv.ValueType = core.ValueInt64
+		if v, ok := dp.Value.(int64); ok {
+			mv.IntValue = int64(v)
+		}
+	case metrics.Gauge:
+		mv.ValueType = core.ValueFloat
+		f, err := metrics.ConvertToFloat64(dp.Value)
+		if err != nil {
+			return tmv, err
+		}
+		mv.FloatValue = float32(f)
+		// if v, ok := dp.Value.(float64); ok {
+
+		// }
+	}
+	tmv.MetricValue = mv
+
+	return tmv, nil
+}
+
+func (h *hawkularSink) metricNameToHawkularType(metricName string) metrics.MetricType {
+	for _, metric := range core.AllMetrics {
+		if metric.Name == metricName {
+			return heapsterTypeToHawkularType(metric.Type)
+		}
+	}
+	return metrics.Gauge
+}
+
+func (h *hawkularSink) keyToTags(metricKey *core.HistoricalKey) map[string]string {
+	tags := make(map[string]string)
+	tags[core.LabelMetricSetType.Key] = metricKey.ObjectType
+
+	switch metricKey.ObjectType {
+	case core.MetricSetTypeSystemContainer:
+		tags[core.LabelNodename.Key] = metricKey.NodeName
+		tags[core.LabelContainerName.Key] = metricKey.ContainerName
+	case core.MetricSetTypePodContainer:
+		tags[core.LabelContainerName.Key] = metricKey.ContainerName
+
+		if metricKey.PodId != "" {
+			tags[core.LabelPodId.Key] = metricKey.PodId
+		} else {
+			tags[core.LabelPodName.Key] = metricKey.PodName
+			tags[core.LabelNamespaceName.Key] = metricKey.NamespaceName
+		}
+	case core.MetricSetTypePod:
+		if metricKey.PodId != "" {
+			tags[core.LabelPodId.Key] = metricKey.PodId
+		} else {
+			tags[core.LabelNamespaceName.Key] = metricKey.NamespaceName
+			tags[core.LabelPodName.Key] = metricKey.PodName
+		}
+	case core.MetricSetTypeNamespace:
+		tags[core.LabelNamespaceName.Key] = metricKey.NamespaceName
+	case core.MetricSetTypeNode:
+		tags[core.LabelNodename.Key] = metricKey.NodeName
+	case core.MetricSetTypeCluster:
+		// System tenant?
```

Comment:
An extra question mark ;). It's just a reminder that "TypeCluster" must always target the default tenant in Openshift (that's where the cluster level information is stored in)